solver/src/settlement_submission/retry.rs

```diff
@@ -0,0 +1,136 @@
+use super::{EncodedSettlement, GAS_PRICE_REFRESH_INTERVAL, MAX_GAS};
+use anyhow::Result;
+use contracts::GPv2Settlement;
+use ethcontract::{
+    errors::{ExecutionError, MethodError},
+    jsonrpc::types::Error as RpcError,
+    transaction::{confirm::ConfirmParams, ResolveCondition},
+    web3::error::Error as Web3Error,
+    GasPrice,
+};
+use futures::{stream, Stream, StreamExt};
+use gas_estimation::GasPriceEstimating;
+use primitive_types::U256;
+use std::time::Duration;
+use transaction_retry::{TransactionResult, TransactionSending};
+
+// Create a never ending stream of gas prices based on checking the estimator in fixed intervals
+// and enforcing the minimum increase. Errors are ignored.
+pub fn gas_price_stream(
+    target_confirm_time: Duration,
+    gas_price_cap: f64,
+    estimator: &dyn GasPriceEstimating,
+) -> impl Stream<Item = f64> + '_ {
+    let stream = stream::unfold(true, move |first_call| async move {
+        if !first_call {
+            tokio::time::delay_for(GAS_PRICE_REFRESH_INTERVAL).await;
+        }
+        let estimate = estimator
+            .estimate_with_limits(MAX_GAS as f64, target_confirm_time)
+            .await;
+        Some((estimate, false))
+    })
+    .filter_map(|gas_price_result| async move {
+        match gas_price_result {
+            Ok(gas_price) => {
+                tracing::debug!("estimated gas price {}", gas_price);
+                Some(gas_price)
+            }
+            Err(err) => {
+                tracing::error!("gas price estimation failed: {:?}", err);
+                None
+            }
+        }
+    });
+    transaction_retry::gas_price_increase::enforce_minimum_increase_and_cap(gas_price_cap, stream)
+}
+
+fn is_transaction_error(error: &ExecutionError) -> bool {
```

Comment:
It's a shame we are duplicating this logic. Would it not make sense to also expose this method in the retry crate (not blocking this PR but generally)?

Reply:
I don't think it makes sense in the retry crate because it is ethcontract specific. Maybe behind an ethcontract feature but then it is annoying when the version of ethcontract doesn't match. Maybe it would make sense in ethcontract or web3 directly but I don't think so because it is so node specific.